src/main/java/com/streamr/client/subs/HistoricalSubscription.java

```diff
@@ -2,41 +2,94 @@
+                keys.add(groupKeys.get(publisherId));
+                keySequences.put(publisherId, new DecryptionKeySequence(keys));
+            }
+        }
     }
 
-    public HistoricalSubscription(String streamId, int partition, MessageHandler handler, ResendOption resendOption, Map<String, String> groupKeysHex,
+    public HistoricalSubscription(String streamId, int partition, MessageHandler handler, ResendOption resendOption,
+                                  Map<String, GroupKey> groupKeys, GroupKeyRequestFunction groupKeyRequestFunction,
                                   long propagationTimeout, long resendTimeout) {
-        this(streamId, partition, handler, resendOption, groupKeysHex, propagationTimeout, resendTimeout, null);
+        this(streamId, partition, handler, resendOption, groupKeys, groupKeyRequestFunction, propagationTimeout, resendTimeout, null);
     }
 
-    public HistoricalSubscription(String streamId, int partition, MessageHandler handler, ResendOption resendOption, Map<String, String> groupKeysHex) {
-        super(streamId, partition, handler, groupKeysHex);
-        this.resendOption = resendOption;
+    public HistoricalSubscription(String streamId, int partition, MessageHandler handler, ResendOption resendOption,
+                                  Map<String, GroupKey> groupKeys, GroupKeyRequestFunction groupKeyRequestFunction) {
+        this(streamId, partition, handler, resendOption, groupKeys, groupKeyRequestFunction,
+                Subscription.DEFAULT_PROPAGATION_TIMEOUT, Subscription.DEFAULT_RESEND_TIMEOUT);
+    }
+
+    public HistoricalSubscription(String streamId, int partition, MessageHandler handler, ResendOption resendOption, Map<String, GroupKey> groupKeys) {
+        this(streamId, partition, handler, resendOption, groupKeys, null);
     }
 
     public HistoricalSubscription(String streamId, int partition, MessageHandler handler, ResendOption resendOption) {
-        super(streamId, partition, handler);
-        this.resendOption = resendOption;
+        this(streamId, partition, handler, resendOption, null);
     }
 
-    public HistoricalSubscription(String streamId, int partition, MessageHandler handler) {
-        super(streamId, partition, handler);
+    @Override
+    public void setGroupKeys(String publisherId, ArrayList<GroupKey> groupKeys) {
+        if (keySequences.containsKey(publisherId)) {
+            throw new InvalidGroupKeyResponseException("Received historical keys for publisher " + publisherId + " for a second time.");
+        }
+        keySequences.put(publisherId, new DecryptionKeySequence(groupKeys));
+        // handle the historical messages received while we were waiting for the historical group keys
+        waitingForGroupKey.remove(publisherId);
+        while (!encryptedMsgsQueue.isEmpty()) {
+            handleInOrder(encryptedMsgsQueue.poll());
+        }
+        if (resendDone) { // the messages in the queue were the last ones to handle
```

Comment:
This code is triggered upon reception of a `ResendResponseResent` or a `ResendResponseNoResend`.
If there still were messages in the `OrderingUtil`, this would mean that a detected gap prevented them to be processed --> there would be some missing messages --> the backend shouldn't have responded with a `ResendResponseResent`. 
This is probably out of scope of this PR, since this question could be asked even without encryption/decryption. It makes things much easier if we can trust the backend when receiving a `ResendResponseResent` but I don't know if that's a reasonable assumption.

Reply:
I think currently the broker (backend) sends a `ResendResponseResent` after it has read all relevant data from Cassandra and pushed it to the websocket connection. It doesn't  check for gaps or anything like that. If there are gaps, there isn't much it can't do since those messages would be more or less lost. But I think this behaviour may be a problem now?